### DISCUSSION TITLE
Best strategy to handle multitenant app with column case differences (postgres/oracle)

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#9538] [Usage Questions] [answered]
Hello,
I am developing an application that works with multiple existing databases.
In my dataset, the columns name respect the default case of each database vendor.
Ex postgres: lowercase, oracle: uppercase.

The select queries work well as long as the identifiers are not quoted. For specific columns, ex `_12col` or some keywords, the column identifier must be quoted and the case must be adapted.
I don't know where I could adapt the case in my declarative model.

Is it possible to handle such a case with a declarative model?
Could you propose me a better strategy?

I have thought about mixing automapped database with declarative tables, I don't know if it's a good solution.

Thanks.

```python
Base = declarative_base()

class Item(Base):
    __tablename__ = "item"
    id = Column(Integer, Identity(), primary_key=True)
    name = Column(String(50))
    _12col = Column(String(50)) # name="_12col" or "_12COL" depending on the dialect (could be set per connection, there may be exceptions)

session.query(item._12col)
```

**Comment by zzzeek** [comment]
There's a bunch of ways I can see this happening, mostly at the compiler / execute level.

Is the "postgresql" / "oracle" thing something you know when the application starts up, and can be a global variable?  that would be the only way this would be an ORM / declarative kind of problem.  Otherwise I think things have to happen within compilers / statement processors.

**Comment by nmoreaud** [comment, marked as answer]
No, my app connects to many databases (documents) which share a common design. It can connect to both Oracle and postgres at the same time.

I could wrap the table classes depending on the dialect, but it would make query code more verbose.
Something like: 
```python
def base_for_dialect(dialect):
  Base = declarative_base()

  class Item(Base):
      __tablename__ = "item"
      id = Column(Integer, Identity(), primary_key=True)
      name = Column(String(50))
      _12col = Column(String(50)) # name="_12col" or "_12COL" depending on the dialect (could be set per connection, there may be 
  exceptions)

  Base.metadata.tables...columns...name = uppercase()

Base = base_for_dialect('oracle')
connection.t = [base. all tables]

session.query(connection.t.Item._12col)  # Model class is declared multiple times
```

**Reply by zzzeek** [reply]
Continuing on my previous comment, this really should be done at the compiler level where the string is set up to be whatever it will be inside the SQL statement.  Managing separate Table objects will be very complicated.

There's many ways to do this, but the one that I think of first is that we can assign these particular strings that we know must be quoted with a special marker around the usual quoted_name construct, and inject a custom quoting scheme into the compilers.   IMO this is definitely the cleanest way to do this.   There's no private API involved, it uses well documented dialect hooks that all the third party dialects have to code towards, so this should remain fairly solid.  POC is below

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Identity
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql.elements import quoted_name

Base = declarative_base()



class ThingString(quoted_name):
    """Some kind of 'marker' string we can search for when identifiers
    are being prepared."""

    def __new__(cls, value: str):
        self = super().__new__(cls, value, quote=True)
        return self


class Item(Base):
    __tablename__ = "model"
    id = Column(Integer, Identity(), primary_key=True)
    name = Column(String(50))
    _12col = Column(ThingString("_12col"), String(50))

e1 = create_engine("postgresql+psycopg2://")
e2 = create_engine("oracle+cx_oracle://")

# dialect.preparer is a class.  Pulling it from a the constructed Dialect on an Engine is a shortcut
# to giving us the exact class for that dialect + driver
# (usually just dialect specific), but could also import the class from the dialects specifically
class PGIdent(e1.dialect.preparer):
    def quote(self, ident, force = None):
        if isinstance(ident, ThingString):
            return super().quote(str(ident).lower())
        return super().quote(ident)

class OracleIdent(e2.dialect.preparer):
    def quote(self, ident, force = None):
        if isinstance(ident, ThingString):
            return super().quote(str(ident).upper())
        return super().quote(ident)

# assign it to an existing dialect, which would already have assigned
# "identifier_preparer" on construction.
e1.dialect.identifier_preparer = PGIdent(e1.dialect)
e2.dialect.identifier_preparer = OracleIdent(e2.dialect)


stmt = select(Item).where(Item._12col == 'hi')

print(stmt.compile(e1))
print(stmt.compile(e2))
```

**Comment by nmoreaud** [comment]
It seems to work well, thanks very much
